src/signature.c

```diff
@@ -171,6 +171,69 @@ GBytes *cms_sign(GBytes *content, const gchar *certfile, const gchar *keyfile, G
 	return res;
 }
 
+gchar* cert_to_text(X509 *cert) {
+	BIO *mem;
+	gchar *data;
+
+	mem = BIO_new(BIO_s_mem());
+	X509_print_ex(mem, cert, 0, 0);
+
+	BIO_get_mem_data(mem, &data);
+
+	BIO_set_close(mem, BIO_NOCLOSE); /* So BIO_free() leaves BUF_MEM alone */
+	BIO_free(mem);
+
+	return data;
+}
+
+gchar* cert_chain_to_text(STACK_OF(X509) *verified_chain) {
+	GString *text = g_string_new(NULL);
+	char buf[BUFSIZ];
+
+	g_string_append(text, "Certificate Chain:\n");
+	for (int i = 0; i < sk_X509_num(verified_chain); i++) {
+		X509_NAME_oneline(X509_get_subject_name(sk_X509_value(verified_chain, i)),
+				buf, sizeof buf);
+		g_string_append_printf(text, "%2d s:%s\n", i, buf);
+		X509_NAME_oneline(X509_get_issuer_name(sk_X509_value(verified_chain, i)),
+				buf, sizeof buf);
+		g_string_append_printf(text, "   i:%s\n", buf);
+	}
+
+	return g_string_free(text, FALSE);
+}
+
+X509 *cms_get_signer_info(CMS_ContentInfo *cms, X509_STORE *store, STACK_OF(X509) **verified_chain) {
+	STACK_OF(X509) *signers = NULL;
+	X509 *signer = NULL;
+
+	signers = CMS_get0_signers(cms);
+
+	g_debug("Got %d signers:", sk_X509_num(signers));
+
+	if (verified_chain) {
+		X509_STORE_CTX *cert_ctx;
+
+		cert_ctx = X509_STORE_CTX_new();
+		if (cert_ctx == NULL)
+			goto out;
+		if (!X509_STORE_CTX_init(cert_ctx, store, sk_X509_value(signers, 0), NULL))
```

Comment:
If I remember correctly, the signature may contain additional certificates for completing the chain. Do we need to pass them here via the chain argument?

Reply:
It is possible to include the certs in the signature, but we don't do that, yet. When required, we should implement it both for creation and verification in one series.